math/cmath/include/algebra/math/impl/cmath_operator.hpp

```diff
@@ -113,4 +114,201 @@ ALGEBRA_HOST_DEVICE inline array_t<scalar_t, 3> operator+(
+      ret[j][i] = a[j][i] * static_cast<scalar_t>(s);
+    }
+  }
+
+  return ret;
+}
+
+template <typename size_type, template <typename, size_type> class array_t,
+          typename scalar_t, size_type ROWS, size_type COLS>
+ALGEBRA_HOST_DEVICE inline array_t<array_t<scalar_t, ROWS>, COLS> operator*(
+    const array_t<array_t<scalar_t, ROWS>, COLS> &a, double s) {
+
+  array_t<array_t<scalar_t, ROWS>, COLS> ret;
+
+  for (size_type i = 0; i < ROWS; ++i) {
+    for (size_type j = 0; j < COLS; ++j) {
+      ret[j][i] = a[j][i] * static_cast<scalar_t>(s);
+    }
+  }
+
+  return ret;
+}
+
+template <typename size_type, template <typename, size_type> class array_t,
+          typename scalar_t, size_type ROWS, size_type COLS>
+ALGEBRA_HOST_DEVICE inline array_t<array_t<scalar_t, ROWS>, COLS> operator*(
+    float s, const array_t<array_t<scalar_t, ROWS>, COLS> &a) {
+
+  return a * s;
+}
+
+template <typename size_type, template <typename, size_type> class array_t,
+          typename scalar_t, size_type ROWS, size_type COLS>
+ALGEBRA_HOST_DEVICE inline array_t<array_t<scalar_t, ROWS>, COLS> operator*(
+    double s, const array_t<array_t<scalar_t, ROWS>, COLS> &a) {
+
+  return a * s;
+}
+
+template <typename size_type, template <typename, size_type> class array_t,
+          typename scalar_t, size_type M, size_type N, size_type O>
+ALGEBRA_HOST_DEVICE inline array_t<array_t<scalar_t, M>, O> operator*(
+    const array_t<array_t<scalar_t, M>, N> &A,
+    const array_t<array_t<scalar_t, N>, O> &B) {
+
+  array_t<array_t<scalar_t, M>, O> C;
+
+  for (size_type i = 0; i < M; ++i) {
+    for (size_type j = 0; j < O; ++j) {
+
+      scalar_t val = 0;
+
+      for (size_type k = 0; k < N; ++k) {
+        val += A[k][i] * B[j][k];
+      }
+
+      C[j][i] = val;
+    }
+  }
+
+  return C;
+}
+
+template <typename size_type, template <typename, size_type> class array_t,
+          typename scalar_t, size_type ROWS, size_type COLS>
+ALGEBRA_HOST_DEVICE inline array_t<array_t<scalar_t, ROWS>, COLS> operator+(
+    const array_t<array_t<scalar_t, ROWS>, COLS> &A,
+    const array_t<array_t<scalar_t, ROWS>, COLS> &B) {
+
+  array_t<array_t<scalar_t, ROWS>, COLS> C;
+
+  for (size_type i = 0; i < ROWS; ++i) {
+    for (size_type j = 0; j < COLS; ++j) {
+      C[j][i] = A[j][i] + B[j][i];
+    }
+  }
+
+  return C;
+}
+
+template <typename size_type, template <typename, size_type> class array_t,
+          typename scalar_t, size_type ROWS, size_type COLS>
+ALGEBRA_HOST_DEVICE inline array_t<array_t<scalar_t, ROWS>, COLS> operator-(
+    const array_t<array_t<scalar_t, ROWS>, COLS> &A,
+    const array_t<array_t<scalar_t, ROWS>, COLS> &B) {
+
+  array_t<array_t<scalar_t, ROWS>, COLS> C;
+
+  for (size_type i = 0; i < ROWS; ++i) {
+    for (size_type j = 0; j < COLS; ++j) {
+      C[j][i] = A[j][i] - B[j][i];
+    }
+  }
+
+  return C;
+}
+
+template <typename size_type, template <typename, size_type> class array_t,
```

Comment:
As I wrote earlier, I would not declare these functions alongside the operators.

The operators are fine like this, and they should be pulled into the global namespace by the front-end header. But these "more complicated" functions should rather stay in the `algebra` namespace. And I would probably not want to define them in this header either, and would only have actual operators here.